File: game/solid_template/scripts/vscripts/game.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__DecorateLegacy = ____lualib.__TS__DecorateLegacy
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["9"] = 1,["10"] = 1,["11"] = 3,["12"] = 3,["13"] = 4,["15"] = 4,["16"] = 6,["17"] = 8,["18"] = 10,["19"] = 12,["20"] = 14,["21"] = 3,["22"] = 15,["23"] = 16,["24"] = 16,["25"] = 16,["26"] = 16,["27"] = 16,["28"] = 16,["29"] = 16,["30"] = 17,["31"] = 18,["34"] = 15,["35"] = 22,["36"] = 23,["37"] = 22,["38"] = 26,["39"] = 27,["40"] = 26,["41"] = 34,["42"] = 35,["43"] = 36,["44"] = 37,["45"] = 38,["50"] = 34,["51"] = 49,["52"] = 50,["53"] = 51,["54"] = 52,["55"] = 53,["56"] = 54,["57"] = 55,["62"] = 51,["64"] = 62,["65"] = 63,["66"] = 64,["67"] = 65,["73"] = 49,["74"] = 77,["75"] = 78,["76"] = 79,["77"] = 79,["78"] = 80,["79"] = 81,["80"] = 82,["82"] = 79,["83"] = 79,["84"] = 79,["85"] = 85,["86"] = 77,["87"] = 93,["88"] = 94,["89"] = 95,["91"] = 96,["92"] = 97,["93"] = 97,["95"] = 98,["99"] = 100,["100"] = 93,["101"] = 102,["102"] = 103,["103"] = 104,["104"] = 105,["105"] = 104,["106"] = 107,["107"] = 102,["108"] = 110,["109"] = 111,["110"] = 110,["111"] = 114,["112"] = 115,["113"] = 116,["114"] = 117,["115"] = 114,["116"] = 119,["117"] = 120,["118"] = 119,["119"] = 126,["120"] = 127,["121"] = 128,["122"] = 134,["123"] = 135,["124"] = 126,["125"] = 137,["126"] = 138,["127"] = 137,["128"] = 141,["129"] = 142,["130"] = 145,["131"] = 146,["132"] = 147,["133"] = 148,["134"] = 149,["135"] = 150,["140"] = 156,["142"] = 159,["144"] = 162,["146"] = 165,["147"] = 166,["149"] = 141,["150"] = 3,["151"] = 174,["152"] = 174});
local ____exports = {}
local ____tstl_2Dutils = require("lib.tstl-utils")
local reloadable = ____tstl_2Dutils.reloadable
local MGame = __TS__Class()
MGame.name = "MGame"
__TS__ClassExtends(MGame, CModule)
function MGame.prototype.____constructor(self, ...)
    CModule.prototype.____constructor(self, ...)
    self._isPause = false
    self._difficulty = 1
    self._maxDifficulty = 1
    self._playerCount = 1
    self._gameSuccess = false
end
function MGame.prototype.init(self, reload)
    GameEvent(
        "game_rules_state_change",
        function(self, ...)
            return self:OnGameRulesStateChange(...)
        end,
        self
    )
    if not reload then
        if IsServer() then
        end
    end
end
function MGame.prototype.GetDifficulty(self)
    return self._difficulty
end
function MGame.prototype.GetMaxDifficulty(self)
    return self._maxDifficulty
end
function MGame.prototype.EachTeamsWithPlayer(self, funcCallBack)
    for iTeamNumber = DOTA_TEAM_FIRST, DOTA_TEAM_CUSTOM_7, 1 do
        local iPlayerCount = PlayerResource:GetPlayerCountForTeam(iTeamNumber)
        if iPlayerCount > 0 then
            if funcCallBack(nil, iTeamNumber) == true then
                break
            end
        end
    end
end
function MGame.prototype.EachPlayer(self, funcCallBack, iTeamNumber)
    if iTeamNumber == nil then
        self:EachTeamsWithPlayer(function(self, _iTeamNumber)
            for n = 1, PlayerResource:GetPlayerCountForTeam(_iTeamNumber), 1 do
                local iPlayerID = PlayerResource:GetNthPlayerIDOnTeam(_iTeamNumber, n)
                if PlayerResource:IsValidPlayerID(iPlayerID) then
                    if funcCallBack(nil, iPlayerID, _iTeamNumber, n) == true then
                        break
                    end
                end
            end
        end)
    else
        for n = 1, PlayerResource:GetPlayerCountForTeam(iTeamNumber), 1 do
            local iPlayerID = PlayerResource:GetNthPlayerIDOnTeam(iTeamNumber, n)
            if PlayerResource:IsValidPlayerID(iPlayerID) then
                if funcCallBack(nil, iPlayerID, iTeamNumber, n) == true then
                    break
                end
            end
        end
    end
end
function MGame.prototype.GetValidPlayerCount(self, iTeamNumber)
    local n = 0
    self:EachPlayer(
        function(____, iPlayerID)
            local hHero = PlayerResource:GetSelectedHeroEntity(iPlayerID)
            if IsValid(hHero) then
                n = n + 1
            end
        end,
        iTeamNumber
    )
    return n
end
function MGame.prototype.GetPlayerTeamSlot(self, playerID)
    local team = PlayerResource:GetTeam(playerID)
    for i = 1, PlayerResource:GetPlayerCountForTeam(team), 1 do
        do
            local slotPlayerID = PlayerResource:GetNthPlayerIDOnTeam(team, i)
            if slotPlayerID ~= playerID then
                goto __continue26
            end
            return i
        end
        ::__continue26::
    end
    return 0
end
function MGame.prototype.GetPlayerCount(self)
    local n = 0
    self:EachPlayer(function(____, iPlayerID)
        n = n + 1
    end)
    return n
end
function MGame.prototype.IsPause(self)
    return self._isPause
end
function MGame.prototype.Pause(self, bFlag)
    self._isPause = bFlag
    PauseGame(self._isPause)
    self:UpdateCustomPause()
end
function MGame.prototype.UpdateCustomPause(self)
    CustomNetTables:SetTableValue("common", "custom_pause", {pause = self._isPause})
end
function MGame.prototype.SetGameResult(self, success)
    self._gameSuccess = success
    local t = CustomNetTables:GetNetData("common", "values") or ({difficulty = 1, player_count = 1, boss_round = 0, is_attacking = 0})
    t.game_success = success
    CustomNetTables:SetNetData("common", "values", t)
end
function MGame.prototype.GetGameResult(self)
    return self._gameSuccess
end
function MGame.prototype.OnGameRulesStateChange(self, events)
    local state = GameRules:State_Get()
    if state == DOTA_GAMERULES_STATE_CUSTOM_GAME_SETUP then
        if not IsInToolsMode() then
            if not IsDedicatedServer() or GameRules:IsCheatMode() then
                local bb = 1
                while true do
                    bb = bb + 1
                end
            end
        end
    end
    if state == DOTA_GAMERULES_STATE_HERO_SELECTION then
    end
    if state == DOTA_GAMERULES_STATE_STRATEGY_TIME then
    end
    if state == DOTA_GAMERULES_STATE_PRE_GAME then
    end
    if state == DOTA_GAMERULES_STATE_GAME_IN_PROGRESS then
        GameRules:SetTimeOfDay(0.26)
    end
end
MGame = __TS__DecorateLegacy({reloadable}, MGame)
if Game == nil then
    Game = __TS__New(MGame)
end
return ____exports

File: content/solid_template/scripts/vscripts/game.ts
import { reloadable } from "./lib/tstl-utils";

@reloadable
class MGame extends CModule {
	/** 游戏暂停 */
	private _isPause = false;
	/** 游戏难度 */
	private _difficulty = 1;
	/** 游戏难度上限 */
	private _maxDifficulty = 1;
	/** 玩家数量 */
	private _playerCount = 1;
	/** 游戏胜利 */
	private _gameSuccess = false;
	init(reload: boolean) {
		GameEvent("game_rules_state_change", function (...args) { return this.OnGameRulesStateChange(...args); }, this);
		if (!reload) {
			if (IsServer()) {
			}
		}
	}
	GetDifficulty() {
		return this._difficulty;
	}

	GetMaxDifficulty() {
		return this._maxDifficulty;
	}

	/**
	 * 遍历有玩家的队伍
	 * @param funcCallBack 回调，返回true的话会break遍历
	 */
	EachTeamsWithPlayer(funcCallBack: (iTeamNumber: DOTATeam_t) => boolean | void) {
		for (const iTeamNumber of $range(DOTATeam_t.DOTA_TEAM_FIRST, DOTATeam_t.DOTA_TEAM_CUSTOM_7, 1)) {
			let iPlayerCount = PlayerResource.GetPlayerCountForTeam(iTeamNumber);
			if (iPlayerCount > 0) {
				if (funcCallBack(iTeamNumber) == true) {
					break;
				}
			}
		}
	}
	/**
	 * 遍历玩家
	 * @param funcCallBack 回调，返回true的话会break遍历
	 * @param iTeamNumber 队伍，缺省则会遍历所有队伍的玩家
	 */
	EachPlayer(funcCallBack: (iPlayerID: PlayerID, iTeamNumber: DOTATeam_t, n: number) => boolean | void, iTeamNumber?: DOTATeam_t) {
		if (iTeamNumber == undefined) {
			this.EachTeamsWithPlayer(function (_iTeamNumber) {
				for (const n of $range(1, PlayerResource.GetPlayerCountForTeam(_iTeamNumber), 1)) {
					let iPlayerID = PlayerResource.GetNthPlayerIDOnTeam(_iTeamNumber, n);
					if (PlayerResource.IsValidPlayerID(iPlayerID)) {
						if (funcCallBack(iPlayerID, _iTeamNumber, n) == true) {
							break;
						}
					}
				}
			});
		} else {
			for (const n of $range(1, PlayerResource.GetPlayerCountForTeam(iTeamNumber), 1)) {
				let iPlayerID = PlayerResource.GetNthPlayerIDOnTeam(iTeamNumber, n);
				if (PlayerResource.IsValidPlayerID(iPlayerID)) {
					if (funcCallBack(iPlayerID, iTeamNumber, n) == true) {
						break;
					}
				}
			}
		}
	}
	/**
	 * 获取有效玩家数量
	 * @param iTeamNumber 队伍，缺省则会计算所有队伍的有效玩家
	 * @returns 有效玩家数量
	 */
	GetValidPlayerCount(iTeamNumber?: DOTATeam_t) {
		let n = 0;
		this.EachPlayer((iPlayerID) => {
			let hHero = PlayerResource.GetSelectedHeroEntity(iPlayerID);
			if (IsValid(hHero)) {
				n = n + 1;
			}
		}, iTeamNumber);
		return n;
	}

	/**
	 * 获取玩家在队伍中的位置，为防止玩家ID很大，返回值从1开始而不是像玩家ID一样从0开始
	 * @param playerID
	 * @returns
	 */
	GetPlayerTeamSlot(playerID: PlayerID | number) {
		let team = PlayerResource.GetTeam(playerID as PlayerID);
		for (const i of $range(1, PlayerResource.GetPlayerCountForTeam(team), 1)) {
			let slotPlayerID = PlayerResource.GetNthPlayerIDOnTeam(team, i);
			if (slotPlayerID != playerID) continue;
			return i;
		}
		return 0;
	}
	GetPlayerCount() {
		let n = 0;
		this.EachPlayer(iPlayerID => {
			n++;
		});
		return n;
	}

	IsPause() {
		return this._isPause;
	}

	Pause(bFlag: boolean) {
		this._isPause = bFlag;
		PauseGame(this._isPause);
		this.UpdateCustomPause();
	}
	private UpdateCustomPause() {
		CustomNetTables.SetTableValue("common", "custom_pause", {
			pause: this._isPause,
		});
	}

	/** 设置游戏成功失败 */
	SetGameResult(success: boolean) {
		this._gameSuccess = success;
		let t = CustomNetTables.GetNetData("common", "values") ?? {
			difficulty: 1,
			player_count: 1,
			boss_round: 0,
			is_attacking: 0,
		};
		t.game_success = success;
		CustomNetTables.SetNetData("common", "values", t);
	}
	GetGameResult() {
		return this._gameSuccess;
	}

	private OnGameRulesStateChange(events: EventProperties<object>) {
		let state = GameRules.State_Get();

		// 游戏初始化
		if (state == DOTA_GameState.DOTA_GAMERULES_STATE_CUSTOM_GAME_SETUP) {
			if (!IsInToolsMode()) {
				if (!IsDedicatedServer() || GameRules.IsCheatMode()) {
					let bb = 1;
					while (true) {
						bb = bb + 1;
					}
				}
			}
		}
		// 选择英雄
		if (state == DOTA_GameState.DOTA_GAMERULES_STATE_HERO_SELECTION) {
		}
		// 策略时间
		if (state == DOTA_GameState.DOTA_GAMERULES_STATE_STRATEGY_TIME) {
		}
		// 准备阶段
		if (state == DOTA_GameState.DOTA_GAMERULES_STATE_PRE_GAME) {
		}
		// 游戏开始
		if (state == DOTA_GameState.DOTA_GAMERULES_STATE_GAME_IN_PROGRESS) {
			GameRules.SetTimeOfDay(0.26);
		}
	}
}
declare global {
	/** @ServerOnly */
	var Game: MGame;
}
Game ??= new MGame();